doc/sphinx/proof-engine/tactics.rst

```diff
@@ -3,35 +3,46 @@
 Tactics
 ========
 
-A deduction rule is a link between some (unique) formula, that we call
-the *conclusion* and (several) formulas that we call the *premises*. A
-deduction rule can be read in two ways. The first one says: “if I know
-this and this then I can deduce this”. For instance, if I have a proof
-of A and a proof of B then I have a proof of A ∧ B. This is forward
-reasoning from premises to conclusion. The other way says: “to prove
-this I have to prove this and this”. For instance, to prove A ∧ B, I
-have to prove A and I have to prove B. This is backward reasoning from
-conclusion to premises. We say that the conclusion is the *goal* to
-prove and premises are the *subgoals*. The tactics implement *backward
-reasoning*. When applied to a goal, a tactic replaces this goal with
-the subgoals it generates. We say that a tactic reduces a goal to its
-subgoal(s).
-
-Each (sub)goal is denoted with a number. The current goal is numbered
-1. By default, a tactic is applied to the current goal, but one can
-address a particular goal in the list by writing n:tactic which means
-“apply tactic tactic to goal number n”. We can show the list of
-subgoals by typing Show (see Section :ref:`requestinginformation`).
-
-Since not every rule applies to a given statement, not every tactic can
-be used to reduce a given goal. In other words, before applying a tactic
-to a given goal, the system checks that some *preconditions* are
-satisfied. If it is not the case, the tactic raises an error message.
-
-Tactics are built from atomic tactics and tactic expressions (which
-extends the folklore notion of tactical) to combine those atomic
-tactics. This chapter is devoted to atomic tactics. The tactic
-language will be described in Chapter :ref:`ltac`.
+Tactics specify how to transform the :term:`proof state` of an
+incomplete proof to eventually generate a complete proof.
+
+Proofs can be developed in two basic ways: In :gdef:`forward reasoning`,
+the proof begins by proving simple statements that are then combined to prove the
+theorem statement as the last step of the proof. With forward reasoning,
+for example,
+the proof of `A /\\ B` would begin with proofs of `A` and `B`, which are
+then used to prove `A /\\ B`.  Forward reasoning is probably the most common
+approach in human-generated proofs.
+
+In :gdef:`backward reasoning`, the proof begins with the theorem statement
+as the goal, which is then gradually transformed until every subgoal generated
+along the way has been proven.  In this case, the proof of `A /\\ B` begins
+with that formula as the goal.  This can be transformed into two subgoals,
+`A` and `B`, followed by the proofs of `A` and `B`.  Coq and its tactics
+use backward reasoning.
```

Comment:
Great rewrite! Much improved compared to the original.

Can I take advantage of this rewrite to ask you to amend slightly the content? (No obligation, this could also be something I do myself later on.)
My issue with the current content is that Coq and its tactics do not actually only support backward reasoning. It's just the most common form in Coq, but some tactics allow introducing "cuts" (to use a logician jargon) and therefore allow proceeding by forward reasoning. These tactics are mostly `assert`, `specialize` and `apply ... in`. In the proof term, they introduce `let ... in` as shown in the following example of forward reasoning proof:

```coq
Parameter A B C : Prop.
Goal A -> (A -> B) -> (B -> C) -> C.
intros Ha Hab Hbc.
apply Hab in Ha as Hb. (* equivalent to specialize (Hab Ha) as Hb. *)
apply Hbc in Hb as Hc. (* equivalent to specialize (Hbc Hb) as Hc. *)
exact Hc.
Show Proof.
(* fun (Ha : A) (Hab : A -> B) (Hbc : B -> C) => let Hb : B := Hab Ha in let Hc : C := Hbc Hb in Hc *)
Qed.
```

In pen and paper proofs, forward and backward reasoning are generally intertwined (when people do a proof by induction, the way they write it is generally "we prove the goal by induction", which is a backward reasoning style, even if they prove each case using a forward reasoning style). It's the same in Coq but backward reasoning is just more common (because practical) in Coq.

Reply:
I've been defining forward and backward reasoning in a specific way, namely that in forward reasoning, you know that the conclusion at each step in proving a goal is true even if the proof isn't completed, whereas in backward reasoning you don't know if any of the conclusions are true until you fully prove the goal.  Maybe I should be explicit about that.  That seemed a too technical to drop in here.  It probably belongs in a unwritten section explaining how human-written proofs map into Coq (e.g. induction, proof by contradiction).

My definition would consider your example backward reasoning.

In a human generated proof using induction, there's probably a comment saying "Now we prove forall n, P(n) by induction on n", followed by proving P(0) and P(n) -> P(n+1), then afterward combining the subproofs to conclude forall n, P(n).  The formal part is still forward reasoning while the comment stating the goal in informal, present only to make the proof more understandable.

But if it's common to refer to your example as forward reasoning, we should use a different definition.  Maybe use my precise definition but also explain that "forward reasoning" may also refer to these other things.  Or make the definition a bit fuzzier, less categorical.

Reply:
I think the original authors of this text probably agreed with what you wrote and I'm the one differing here. So let's keep things as they are now (you did a pretty good job already) and we can have this discussion on how pen-and-paper proofs map to Coq proofs later on.